src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -701,6 +712,47 @@ static int get_pool_offset(sd_dhcp_server *server, be32_t requested_ip) {
         return be32toh(requested_ip & ~server->netmask) - server->pool_offset;
 }
 
+static int dhcp_server_relay_message(sd_dhcp_server *server, DHCPMessage *message, size_t opt_length) {
+        _cleanup_free_ DHCPPacket *packet = NULL;
+
+        assert(server);
+        assert(message);
+        assert(sd_dhcp_server_is_in_relay_mode(server));
+
+        if (message->op == BOOTREPLY) {
+                log_dhcp_server(server, "(relay agent) BOOTREPLY (0x%x)", be32toh(message->xid));
+                if (message->giaddr != server->address) {
+                        log_dhcp_server(server, "(relay agent)  BOOTREPLY giaddr mismatch, discarding");
+                        return -EBADMSG;
+                }
+
+                int message_type = dhcp_option_parse(message, sizeof(DHCPMessage) + opt_length, NULL, NULL, NULL);
+                if (message_type < 0)
+                        return message_type;
+
+                packet = malloc0(sizeof(DHCPPacket) + opt_length);
```

Comment:
malloc0 here seem unnecessary, since the memory is immediately overwritten by memcpy.

Reply:
It' here to zero DHCPPacket part, which is not copied over.

Reply:
Yeah, that's OK. This is a small structure, so there's no need to microoptimize here.